src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,175 @@
+ * To find the best possible solution, we use a waste metric. The waste metric is defined as the cost to spend the current inputs
+ * now minus the cost to spend the current inputs later, plus the amount exceeding the target value. We search the tree to find the
+ * set of UTXOs which falls within our range and minimizes waste.
+ *
+ * An additional optimization of this algorithm implemented here is a lookahead value which maintains the total value of the UTXO set
+ * of all unexplored UTXOs (i.e. UTXOs that have not yet been included or excluded). This allows us to cut a branch if the remaining
+ * amount is not sufficient to reach our target.
+ *
+ * SelectCoinsBnB Arguments:
+ * const std::vector<CInputCoin>& utxo_pool -> The set of UTXOs that we are choosing from. These UTXOs will be sorted in descending order
+ *                                             by value and the CInputCoins' values are their effective values.
+ * const CAmount& target_value -> This is the value that we want to select. It is the lower bound of the range.
+ * const CAmount& cost_of_change -> This is the cost of creating and spending a change output. This plus target_value is the upper bound
+ *                                  of the range.
+ * std::set<CInputCoin>& out_set -> This is an output parameter for the set of CInputCoins that have been selected.
+ * CAmount& value_ret -> This is an output parameter for the total value of the CInputCoins that were selected.
+ * CAmount& fee_ret -> This is an output parameter for the value of the transaction fees for the CInputCoins that were selected.
+ */
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret, CAmount& fee_ret)
+{
+    out_set.clear();
+    CAmount value_track = 0;
+
+    if (utxo_pool.empty()) {
+        return false;
+    }
+
+    int depth = 0;
+    int remaining_tries = 100000;
+    std::vector<bool> selection; // select the utxo at this index
+    selection.assign(utxo_pool.size(), false);
+    bool done = false;
+    bool backtrack = false;
+
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+
+    // Calculate lookahead
+    CAmount lookahead = 0;
+    for (const CInputCoin& utxo : utxo_pool) {
+        lookahead += utxo.txout.nValue;
+    }
+
+    // Best solution
+    CAmount curr_waste = 0;
+    std::vector<bool> best_selection;
+    CAmount best_waste = MAX_MONEY;
+
+    // Depth First search loop for choosing the UTXOs
+    while (!done)
+    {
+        if (remaining_tries <= 0) { // Too many tries, exit
+            break;
+        } else if (value_track + lookahead < target_value) { // Cannot possibly reach target with the amount remaining in the lookahead.
+            // This also catches if we are at a leaf node and still have not met the target value
+            if (depth == 0) { // At the first utxo, no possible selections, so exit
```

Comment:
This should just be a check after calculating the original value. Has no additional checking power here.

Reply:
Removed. Also compacted the next 2 `else if`s with this one to make one large `else if` with `or`.